File: src/lib/community/rideGroupMap.ts
/**
 * Static meeting pins for Browse / Explore — never a live GPS of a member.
 */

import type { RideGroup } from "@/lib/community/types";
import { canShowMeetingOnExplore } from "@/lib/community/rideGroup";
import { parseMeetingLatLng } from "@/lib/community/placesMerger";

export type GroupMeetPin = {
  group: RideGroup;
  lat: number;
  lng: number;
  label: string;
  placeId: string;
};

export function groupMeetPinsOnExplore(input: {
  groups: RideGroup[];
  memberGroupIds: Iterable<string>;
  now?: Date;
  centerFor?: (
    group: RideGroup
  ) => { lat: number; lng: number } | null | undefined;
}): GroupMeetPin[] {
  const members = new Set(input.memberGroupIds);
  const now = input.now ?? new Date();
  const out: GroupMeetPin[] = [];
  for (const group of input.groups) {
    if (
      !canShowMeetingOnExplore({
        visibility: group.visibility,
        status: group.status,
        startWindowEnd: group.startWindowEnd,
        isMember: members.has(group.id),
        now,
      })
    ) {
      continue;
    }
    const parsed = parseMeetingLatLng(group.meetingPoint);
    const fallback = input.centerFor?.(group);
    const lat = parsed?.lat ?? fallback?.lat;
    const lng = parsed?.lng ?? fallback?.lng;
    if (lat == null || lng == null) continue;
    const label = (parsed?.label ?? group.meetingPoint ?? "").trim();
    out.push({
      group,
      lat,
      lng,
      label: label || group.title,
      placeId: `meet-${group.id}`,
    });
  }
  return out;
}

File: mobile/lib/domain/community/ride_group_map.dart
import 'ride_group.dart';
import 'ride_group_policy.dart';
import 'stimme_pin.dart';

/// Static meeting pin for Browse — never a live GPS of a member.
class GroupMeetPin {
  const GroupMeetPin({
    required this.group,
    required this.lat,
    required this.lng,
    required this.label,
  });

  final RideGroup group;
  final double lat;
  final double lng;
  final String label;

  String get placeId => 'meet-${group.id}';
}

/// Public joinable groups (and own groups) as map pins.
/// Uses meeting coordinates when present, else [centerFor].
List<GroupMeetPin> groupMeetPinsOnExplore({
  required List<RideGroup> groups,
  required Set<String> memberGroupIds,
  required DateTime now,
  ({double lat, double lng})? Function(RideGroup group)? centerFor,
}) {
  final out = <GroupMeetPin>[];
  for (final g in groups) {
    if (!RideGroupPolicy.canShowMeetingOnExplore(
      g,
      isMember: memberGroupIds.contains(g.id),
      now: now,
    )) {
      continue;
    }
    final parsed = parseMeetingLatLng(g.meetingPoint);
    final fallback = centerFor?.call(g);
    final lat = parsed?.lat ?? fallback?.lat;
    final lng = parsed?.lng ?? fallback?.lng;
    if (lat == null || lng == null) continue;
    final label = (parsed?.label ?? g.meetingPoint ?? '').trim();
    out.add(
      GroupMeetPin(
        group: g,
        lat: lat,
        lng: lng,
        label: label.isEmpty ? g.title : label,
      ),
    );
  }
  return out;
}
